src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+                val canonical_body_index = qs.getBodyIndex(varName)
+                if (canonical_body_index != bodyIndex) {
+                  val real_attr_name1 = resolveName( Variable(varName, relName, atomIndex) )
+                  val real_attr_name2 = resolveName( qs.getVar(varName))
+                  Some(s"R${ bodyIndex }.${ real_attr_name1 } = R${ canonical_body_index }.${ real_attr_name2 } ")
+                } else { None }
+              }
+              case Placeholder() => None
+              // other expressions indicate a filter condition on the column
+              case _ => {
+                val resolved = compileExpr(expr, z)
+                val attr = schema(relName, atomIndex)
+                Some(s"R${bodyIndex}.${attr} = ${resolved}")
+              }
             }
-            // other expressions indicate a filter condition on the column
-            case _ => {
-              val resolved = compileExpr(expr, z, OriginalOnly, index, false)
-              val attr = schema(relName, index)
-              Some(s"R${bodyIndex}.${attr} = ${resolved}")
+          }
+        }
+        case (x: Cond, _) => Some(compileCond(x, z))
+        case (x: ModifierAtom, index) => {
+          val newIndexPrefix = s"${indexPrefix}${index}_"
+          val subqueryWhereStr = generateWhereClause(x.bodies, newIndexPrefix)
+          val subqueryFromStr  = generateFromClause(x.bodies, newIndexPrefix)
+          x.modifier match {
+            case ExistModifier(negated) => Some(s"${if (negated) "NOT " else ""}EXISTS (SELECT 1 FROM ${subqueryFromStr} WHERE ${subqueryWhereStr})")
+            case _ => None
+          }
+        }
+      }
+      conditions.mkString(" AND ")
+    }
+
+    def generateFromClause(bodies: List[Body], indexPrefix: String) : String = {
+      val atoms = bodies.zipWithIndex flatMap {
+        case (x:Atom,i) => Some(s"${x.name} R${indexPrefix}${i}")
+        case _ => None
+      }
+      val outers = bodies.zipWithIndex flatMap {
+        case (x:ModifierAtom,i) => {
+          val newIndexPrefix = s"${indexPrefix}${i}_"
+          x.modifier match {
+            case OuterModifier() => {
+              val from = generateFromClause(x.bodies, newIndexPrefix)
+              val joinCond = generateWhereClause(x.bodies, newIndexPrefix)
+              Some(s"${from} ON ${joinCond}")
             }
+            case _ => None
           }
         }
+        case _ => None
+      }
+      // full outer join
+      if (atoms isEmpty) {
+        outers.mkString(" FULL OUTER JOIN ")
+      } else {
+        (atoms.mkString(", ") +: outers).mkString(" LEFT OUTER JOIN ")
       }
     }
 
-    // resolve conditions
-    val conditionStr = z.conditions(0) map (compileCond(_, z))
+    val whereClause = generateWhereClause(z.bodies(0), "")
+    val whereClauseStr = if (whereClause.isEmpty) "" else "WHERE " + whereClause
+    val fromClauseStr = generateFromClause(z.bodies(0), "")
```

Comment:
Can we inline these at the end where the whole SQL query is put together?

I like the structure we're getting at: defining functions that take care of a particular aspect of the compilation, then using them at the end to generate the final result.  So, let's move most `val`s down.

Defining a function that surrounds certain prefix/postfix strings only for non-empty args may be useful:
```scala
def optionalClause(prefix: String, clause: String, postfix: String = ""): String =
  if (clause isEmpty) "" else prefix + clause + postfix
```